xamarin/Xamarin.Forms
# Label padding

### Description of Change ###

Added Padding property to Label.

_WARNING: It might look boring to read the text below, but it's all worth it, I promise! Thanks!_

### iOS behavior and implementation ###
All platforms beside iOS (and MacOS) support padding on their native label, so we already know that the  real issue is with iOS.

We also know this can be easily achieved on iOS by having a UILabel derived class and overriding some of the UILabel methods (which is what I did in this PR too). However, using a derived UILabel to render any Label can be a performance issue on both native iOS and Xamarin. How much it hurts the performance? It depends on how heavy a Page is, how many labels exist on it.

**The compromise for iOS:**
If a Label is created with a non-zero Padding, on iOS the label renderer will render a FormsLabel class derived from UILabel and Padding will work:
```
_myLabel = new Label() { Padding = new Thickness(40, 20) }; 
stackLayout.Add(_myLabel);

// Later on, after Page was rendered:
void OnMyBtnClicked()   
{
   _myLabel.Padding = new Thickness(5);  // This will work, padding is updated
}
```

If Padding is not set (it's zero), it will render to UILabel, like it's working today. In this case, updating the Padding **will be ignored** after the Label is created:
```
_myLabel = new Label(); 
stackLayout.Add(_myLabel);

// Later on, after Page was rendered:
void OnMyBtnClicked() 
{
    _myLabel.Padding = new Thickness(5);  // iOS ONLY: This will NOT work, padding is not set
}
```

In other words, if you need Padding on the Label, set it when you create the Label and it's added to the Page, not afterwards. I think it's a good compromise, because in practice, there are very few scenarios where Padding needs to be updated. 

**What if an app really needs to update the Padding which is initially empty?** 
If this is really needed, one option is to overload the Label c-tor with a boolean parameter to indicate the iOS renderer that Padding will be updated, just how the [`CachingStrategy`](https://docs.microsoft.com/en-us/xamarin/xamarin-forms/user-interface/listview/performance#setting-the-caching-strategy) works on ListView, and it will have a purpose only for iOS:
```
XAML: <Label IsPaddingUpdated="true" />
C#: _myLabel = new Label(isPaddingUpdated: true);
```

Of course, the **3rd option** is to just always render to the UILabel derived class FormsLabel but, like I said above, it can hurt the performance. 

**4th option**: Make the iOS Label renderer switch to the UILabel derived class (FormsLabel) when Padding is set. But I'm not sure, is this is possible at the moment? What are the implications of switching the renderer? If you think it's worth exploring please share input on this.

What do you think?

### Issues Resolved ### 


- fixes #4784

### API Changes ###


- Core/XAML (all platforms)
- iOS
- Android
- UWP
- WPF

### Testing Procedure ###


See the **Label Gallery** page (LabelCoreGalleryPage):

**(IMPORTANT: Note that because of #4143, depending on the place where you tap a Span, tapping might NOT work!!!!!)**

![image](https://user-images.githubusercontent.com/743918/58309592-55c29180-7e0d-11e9-9be2-14cb249ec8de.png)




### PR Checklist ###
